sw/device/lib/dif/dif_padctrl.c

```diff
@@ -0,0 +1,305 @@
+                                             dif_padctrl_attr_t reset_attrs) {
+  if (padctrl == NULL) {
+    return kDifPadctrlResetBadArg;
+  }
+
+  if (!attrs_are_valid(reset_attrs)) {
+    return kDifPadctrlResetBadArg;
+  }
+
+  if (padctrl_is_locked(padctrl)) {
+    return kDifPadctrlResetLocked;
+  }
+
+  // Form attribute register reset value from reptitions of reset_attrs
+  // appropriately shifted and combined.
+  uint32_t attr_reg_reset_val = 0;
+
+  for (int i = 0; i < PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG; ++i) {
+    attr_reg_reset_val <<= PADCTRL_PARAM_ATTRDW;
+    attr_reg_reset_val |= reset_attrs;
+  }
+
+  // Reset all DIO pad attributes to the desired reset value.
+  for (int i = 0; i < PADCTRL_DIO_PADS_MULTIREG_COUNT; ++i) {
+    ptrdiff_t dio_pads_register_offset =
+        PADCTRL_DIO_PADS0_REG_OFFSET + (i * kPadctrlRegisterWidthInBytes);
+    mmio_region_write32(padctrl->base_addr, dio_pads_register_offset,
+                        attr_reg_reset_val);
+  }
+
+  // Reset all MIO pad attributes to the desired reset value.
+  for (int i = 0; i < PADCTRL_MIO_PADS_MULTIREG_COUNT; ++i) {
+    ptrdiff_t mio_pads_register_offset =
+        PADCTRL_MIO_PADS0_REG_OFFSET + (i * kPadctrlRegisterWidthInBytes);
+    mmio_region_write32(padctrl->base_addr, mio_pads_register_offset,
+                        attr_reg_reset_val);
+  }
+
+  return kDifPadctrlResetOk;
+}
+
+dif_padctrl_result_t dif_padctrl_lock(dif_padctrl_t *padctrl) {
+  if (padctrl == NULL) {
+    return kDifPadctrlBadArg;
+  }
+
+  mmio_region_write32(padctrl->base_addr, PADCTRL_REGEN_REG_OFFSET, 0);
+
+  return kDifPadctrlOk;
+}
+
+dif_padctrl_change_attr_result_t dif_padctrl_enable_attrs(
+    const dif_padctrl_t *padctrl, dif_padctrl_pad_kind_t kind,
+    dif_padctrl_pad_index_t index, dif_padctrl_attr_t enable_attrs) {
+  return change_attrs(padctrl, kind, index, enable_attrs, kAttrChangeEnable);
+}
+
+dif_padctrl_change_attr_result_t dif_padctrl_disable_attrs(
+    const dif_padctrl_t *padctrl, dif_padctrl_pad_kind_t kind,
+    dif_padctrl_pad_index_t index, dif_padctrl_attr_t disable_attrs) {
+  return change_attrs(padctrl, kind, index, disable_attrs, kAttrChangeDisable);
+}
+
+dif_padctrl_change_attr_result_t dif_padctrl_write_attrs(
+    const dif_padctrl_t *padctrl, dif_padctrl_pad_kind_t kind,
+    dif_padctrl_pad_index_t index, dif_padctrl_attr_t write_attrs) {
+  return change_attrs(padctrl, kind, index, write_attrs, kAttrChangeWrite);
+}
+
+dif_padctrl_result_t dif_padctrl_get_attrs(const dif_padctrl_t *padctrl,
+                                           dif_padctrl_pad_kind_t kind,
+                                           dif_padctrl_pad_index_t index,
+                                           dif_padctrl_attr_t *attrs_out) {
+  if (padctrl == NULL) {
+    return kDifPadctrlBadArg;
+  }
+
+  if (attrs_out == NULL) {
+    return kDifPadctrlBadArg;
+  }
+
+  if (!is_pad_index_valid(kind, index)) {
+    return kDifPadctrlBadArg;
+  }
```

Comment:
(This applies to the whole change.)

Consider merging preconditions into a single if-statement with `||` connectives.

Reply:
I prefer this style as it makes the condition checks stand-out as individual elements more clearly.